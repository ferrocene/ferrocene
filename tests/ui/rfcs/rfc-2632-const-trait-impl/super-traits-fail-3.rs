#![feature(const_trait_impl)]

// revisions: yy yn ny nn
//[yy] known-bug: #110395
//FIXME [yy] check-pass

#[cfg_attr(any(yy, yn), const_trait)]
trait Foo {
    fn a(&self);
}

#[cfg_attr(any(yy, ny), const_trait)]
trait Bar: ~const Foo {}
//[ny,nn]~^ ERROR: ~const can only be applied to `#[const_trait]`
//[ny,nn]~| ERROR: ~const can only be applied to `#[const_trait]`
//[yn,nn]~^^^ ERROR: `~const` is not allowed here

const fn foo<T: ~const Bar>(x: &T) {
    //[yn,nn]~^ ERROR: ~const can only be applied to `#[const_trait]`
    x.a();
}

fn main() {}
