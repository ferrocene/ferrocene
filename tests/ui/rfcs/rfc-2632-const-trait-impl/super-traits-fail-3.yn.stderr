error: `~const` is not allowed here
  --> $DIR/super-traits-fail-3.rs:14:12
   |
LL | trait Bar: ~const Foo {}
   |            ^^^^^^
   |
note: this trait is not a `#[const_trait]`, so it cannot have `~const` trait bounds
  --> $DIR/super-traits-fail-3.rs:14:1
   |
LL | trait Bar: ~const Foo {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^

error: `~const` can only be applied to `#[const_trait]` traits
  --> $DIR/super-traits-fail-3.rs:20:24
   |
LL | const fn foo<T: ~const Bar>(x: &T) {
   |                        ^^^

error[E0277]: the trait bound `T: ~const Foo` is not satisfied
  --> $DIR/super-traits-fail-3.rs:22:7
   |
LL |     x.a();
   |       ^ the trait `Foo` is not implemented for `T`
   |
note: required by a bound in `Foo::a`
  --> $DIR/super-traits-fail-3.rs:8:25
   |
LL | #[cfg_attr(any(yy, yn), const_trait)]
   |                         ^^^^^^^^^^^ required by this bound in `Foo::a`
LL | trait Foo {
LL |     fn a(&self);
   |        - required by a bound in this associated function
help: consider further restricting this bound
   |
LL | const fn foo<T: ~const Bar + Foo>(x: &T) {
   |                            +++++

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0277`.
